[libc++] P2770R0: "Stashing stashing iterators for proper flattening"

JMazurkiewicz opened the pull request:
* Parially implements [P2770R0: "Stashing stashing iterators for proper flattening"](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p2770r0.html)
* Fixes [LWG3698](https://wg21.link/lwg3698), [LWG3700](https://wg21.link/lwg3700), and [LWG3791](https://wg21.link/lwg3791)
* `join_with_view` hasn't been done yet since this type isn't implemented yet
* Rename `test/libcxx/ranges/range.adaptors/range.adaptor.tuple` directory to `test/libcxx/ranges/range.adaptors/range.adaptor.helpers` to match the standard: http://eel.is/c++draft/range.adaptor.helpers (this change happened in P2770R0, see point 3 of wording).
* Rename `libcxx\test\std\ranges\range.adaptors\range.join.view` to `libcxx\test\std\ranges\range.adaptors\range.join` to match the standard

JMazurkiewicz committed:
[libc++] P2770R0: "Stashing stashing iterators for proper flattening"

* Parially implements P2770R0: "Stashing stashing iterators for proper flattening"
* `join_with_view` hasn't been done yet since this type isn't implemented yet
* Rename `test/libcxx/ranges/range.adaptors/range.adaptor.tuple` directory to `test/libcxx/ranges/range.adaptors/range.adaptor.helpers` to match the standard: http://eel.is/c++draft/range.adaptor.helpers (this change happened in P2770R0, see point 3 of wording).
* Rename `libcxx\test\std\ranges\range.adaptors\range.join.view` to `libcxx\test\std\ranges\range.adaptors\range.join` to match the standard

EricWF commented on libcxx/include/__ranges/join_view.h:
```diff
@@ -84,11 +82,16 @@ namespace ranges {
+
+    static constexpr bool _UseInnerCache = !is_reference_v<_InnerRange>;
+    using _InnerCache = _If<_UseInnerCache, __non_propagating_cache<remove_cvref_t<_InnerRange>>, __empty_cache>;
+    _LIBCPP_NO_UNIQUE_ADDRESS _InnerCache __inner_;
```
I have a preference that all data members of a class are defined next to each other. Could you please move the metaprogramming above both declarations?

JMazurkiewicz replied:
I think I'm going to leave it as it is (for now). Other range adaptors (like `drop_while_view`, `filter_view`, `chunk_by_view`) use similar style.

I can address your comment in follow-up PR to clean up style in other adaptors too.

EricWF commented on libcxx/test/libcxx/ranges/range.adaptors/range.adaptor.helpers/as-lvalue.cpp:
```diff
@@ -0,0 +1,47 @@
+
+  // Check prvalue (reference is dangling)
+  {
+    [[maybe_unused]] std::same_as<int&> decltype(auto) check = std::ranges::__as_lvalue(0);
```
Yeah, this test show exactly how dangerous this API is.

Why do you need it?

JMazurkiewicz replied:
The *`as-lvalue`* function is used by P2770 to fix joining ranges of xvalues (rvalue references), see [LWG-3791](https://cplusplus.github.io/LWG/issue3791) and example: https://godbolt.org/z/Ev17W3b5W.

EricWF commented on libcxx/include/__ranges/helpers.h:
```diff
@@ -0,0 +1,39 @@
+
+namespace ranges {
+template <class _Tp>
+_LIBCPP_HIDE_FROM_ABI constexpr _Tp& __as_lvalue(_Tp&& __t) {
```
I don't like this function, and am very tempted to block the patch on it alone.
I think it's a dangerous way to hide lifetime bugs.

Can you apply the `[[lifetimebound]]` attribute to the parameter and see if it still works?

JMazurkiewicz replied:
> Can you apply the `[[lifetimebound]]` attribute to the parameter and see if it still works?

I've applied it and no error has occurred in `range.join` tests. I've also updated `range.adaptor.helpers/as-lvalue.cpp` test to verify that calling function with prvalue gives a warning.

EricWF reviewed (changes requested):
The __as_lvalue function makes me super nervous, and I would like to have those concerns addressed before moving forward.

JMazurkiewicz commented on libcxx/include/regex:
```diff
@@ -6407,6 +6409,9 @@ public:
     typedef const value_type*                     pointer;
     typedef const value_type&                     reference;
     typedef forward_iterator_tag                  iterator_category;
+#if _LIBCPP_STD_VER >= 20
```
I don't think that this is good idea, `iterator_concept` has no meaning before C++20.

frederick-vs-ja replied:
I don't think we should furtherly backport `iterator_concept` which was not reserved until C++20.

ldionne replied:
Agreed -- we definitely don't want to expose C++20 features or names before C++20.

ldionne commented on libcxx/docs/Status/Cxx23.rst:
```diff
@@ -46,6 +46,7 @@ Paper Status
       clang doesn't issue a diagnostic for deprecated using template declarations.
    .. [#note-P2520R0] P2520R0: Libc++ implemented this paper as a DR in C++20 as well.
    .. [#note-P2711R1] P2711R1: ``join_with_view`` hasn't been done yet since this type isn't implemented yet.
+   .. [#note-P2770R0] P2770R0: Same as the above.
```
I don't understand this note. I'm not sure the note will make sense once this is rendered as HTML?

ldionne commented on libcxx/docs/Status/RangesViews.csv:
```diff
@@ -28,7 +28,7 @@ C++23,`zip <https://wg21.link/P2321R2>`_,Hui Xie,`D122806 <https://llvm.org/D122
 C++23,`zip_transform <https://wg21.link/P2321R2>`_,Hui Xie,No patch yet,Not started
```
Not attached to the diff. We should update `libcxx/docs/UsingLibcxx.rst` to stop mentioning that `std::ranges::join_view` is experimental.

frederick-vs-ja commented on libcxx/include/__ranges/helpers.h:
```diff
@@ -0,0 +1,39 @@
+// -*- C++ -*-
```
> Otherwise it could just live in `join_view.h`.

Hmm, according to the paper, _`as-lvalue`_ should be shared between `join_view` and `join_with_view`, and thus it deseves a separated internal header.

ldionne replied:
Sounds reasonable, what do you think about `__utility/as_lvalue.h`?

JMazurkiewicz replied:
I've moved *`as-lvalue`* to `__utility/as_lvalue.h`, but I've left it in `ranges` namespace since it is used only by `join` and `join_with` views.

JMazurkiewicz commented on libcxx/docs/Status/Cxx23Papers.csv:
```diff
@@ -104,7 +104,7 @@
 "","","","","","",""
 "`P0290R4 <https://wg21.link/P0290R4>`__","LWG", "``apply()`` for ``synchronized_value<T>``","February 2023","","","|concurrency TS|"
-"`P2770R0 <https://wg21.link/P2770R0>`__","LWG", "Stashing stashing ``iterators`` for proper flattening","February 2023","|In Progress|","","|ranges|"
+"`P2770R0 <https://wg21.link/P2770R0>`__","LWG", "Stashing stashing ``iterators`` for proper flattening","February 2023","|Partial| [#note-P2770R0]_","","|ranges|"
```
Yes, but there's no need to mention those issues - they were marked as ["resolved"](https://cplusplus.github.io/LWG/lwg-active.html#Resolved).

ldionne commented on libcxx/test/std/ranges/range.adaptors/range.join.view/iterator/ctor.parent.outer.pass.cpp:
```diff
@@ -1,57 +0,0 @@
-  }
-
-  {
-    // LWG3569 Inner iterator not default_initializable
```
It seems like this part of the test would still be relevant to the following constructor?

```
constexpr iterator(Parent& parent, OuterIter outer)
    requires forward_range<Base>
```

ldionne replied:
Not addressed (unless I missed)?

ldionne commented on libcxx/include/__ranges/join_view.h:
```diff
@@ -234,8 +243,37 @@ namespace ranges {
+      }
+    }
+
+    _LIBCPP_HIDE_FROM_ABI constexpr __iterator(_Parent& __parent, _Outer __outer)
```
Can you please add tests for these constructors? You removed the test in `libcxx/test/std/ranges/range.adaptors/range.join.view/iterator/ctor.parent.outer.pass.cpp` but I don't think you added any for the new ctors. Note that those are exposition only ctors so the tests could go in `libcxx/test/libcxx/<...>` instead of `libcxx/test/std/<...>` -- or if you prefer you can put them in `libcxx/test/std` but mark them with `REQUIRES: stdlib=libc++`.

ldionne replied:
I don't think this has been addressed (unless I missed it)?

ldionne commented on libcxx/test/std/ranges/range.adaptors/range.join/range.join.iterator/eq.pass.cpp:
```diff
@@ -43,16 +42,14 @@ constexpr bool test() {
     std::ranges::join_view jv(Outer{inners});
     auto iter = jv.begin();
     static_assert(!std::equality_comparable<decltype(iter)>);
-    auto const_iter = std::as_const(jv).begin();
```
I don't understand why this isn't valid anymore?

JMazurkiewicz replied:
This is not valid anymore, because `jv` adapts input range (of type `Outer`) and, since P2770R0, `begin() const` requires forward range.

JMazurkiewicz replied:
I've noticed that the comment above is wrong, it should say `!forward_range<Base>` instead of `!forward_range<iterator_t<Base>>`.

ldionne commented on libcxx/test/std/ranges/range.adaptors/range.join/range.join.sentinel/eq.pass.cpp:
```diff
@@ -7,7 +7,6 @@
 //===----------------------------------------------------------------------===//
 
```
Not attached to this line: Can we add somewhere the following example from https://cplusplus.github.io/LWG/issue3698 ?

```
#include <ranges>
#include <regex>
#include <iostream>

int main() {
  char const text[] = "Hello";
  std::regex regex{"[a-z]"};

  auto lower = std::ranges::subrange(
        std::cregex_iterator(
            std::ranges::begin(text),
            std::ranges::end(text),
            regex),
        std::cregex_iterator{}
    )
    | std::views::join
    | std::views::transform([](auto const& sm) {
        return std::string_view(sm.first, sm.second);
    });

  for (auto const& sv : lower) {
    std::cout << sv << '\n';
  }
}
```

Without `cout` of course. I would put that in e.g. `libcxx/test/std/ranges/range.adaptors/range.join/LWG3698.pass.cpp`. You'll need to make that `UNSUPPORTED: no-localization` and perhaps a few others (CI should tell you) since it's using `regex`.

ldionne reviewed (changes requested):
Thanks for the patch and for being patient! I think I did a full review, this should be pretty much good to go once everything has been addressed.

ldionne commented:
By the way, please avoid force-pushing your branch when you update. Instead, create commits on top of your branch and push those. Otherwise, my comments will get detached from the diff and it will be basically impossible to review.

JMazurkiewicz committed:
Enable `__as_lvalue` in >C++03 modes

Comments:
* https://github.com/llvm/llvm-project/pull/66033#discussion_r1396562251
* https://github.com/llvm/llvm-project/pull/66033#discussion_r1396564172
* https://github.com/llvm/llvm-project/pull/66033#discussion_r1396579715

JMazurkiewicz committed:
Split `as-lvalue.verify.cpp`

Comments:
* https://github.com/llvm/llvm-project/pull/66033#discussion_r1396565383
* https://github.com/llvm/llvm-project/pull/66033#discussion_r1396565996

JMazurkiewicz committed:
Add missing `explicit` specifier in `__iterator(_Parent& __parent)`

Comment: https://github.com/llvm/llvm-project/pull/66033#discussion_r1396574310

JMazurkiewicz committed:
Restore old `__sentinel<C>` friendship in `join_view::_sentinel`

Comment: https://github.com/llvm/llvm-project/pull/66033#discussion_r1396587214

JMazurkiewicz committed:
Un-experimental more `join_view` related tests

JMazurkiewicz committed:
Make `join_view::__iterator::__outer_` private

Which is hopefully enough.

Comment: https://github.com/llvm/llvm-project/pull/66033#discussion_r1396575242

JMazurkiewicz committed:
Add test with code from LWG-3698

Comment: https://github.com/llvm/llvm-project/pull/66033#discussion_r1396600533

JMazurkiewicz committed:
Whitespace amendments

JMazurkiewicz committed:
Add extra static assertion for `[range.join.iterator] Note 1`

Comment: https://github.com/llvm/llvm-project/pull/66033#discussion_r1396578912

JMazurkiewicz committed:
Fix comment: `!forward_range<iterator_t<Base>>` -> `!forward_range<Base>`

Comment: https://github.com/llvm/llvm-project/pull/66033#discussion_r1396594058

JMazurkiewicz committed:
Merge branch 'main' into libcxx/ranges/stashing2

JMazurkiewicz committed:
Make `as-lvalue.pass.cpp` C++11 friendly

ldionne commented on libcxx/test/libcxx/ranges/range.adaptors/range.adaptor.helpers/as-lvalue.pass.cpp:
```diff
@@ -0,0 +1,30 @@
+//
+//===----------------------------------------------------------------------===//
+
+// UNSUPPORTED: c++03
```
I'd prefer if we did `UNSUPPORTED: c++03, c++11` so we can use C++14 constexpr. It's nice that you try to make it work in C++11, but TBH I think we should prioritize readability for these tests -- at the end of the day `__as_lvalue` is also not going to be used in C++11 mode so I'm not too shocked about not having coverage for it in that mode.

Then you can use multiple statements in the function below.

ldionne commented on libcxx/test/std/ranges/range.adaptors/range.join/lwg3698.pass.cpp:
```diff
@@ -0,0 +1,31 @@
+          std::cregex_iterator(std::ranges::begin(text), std::ranges::end(text), regex), std::cregex_iterator{}) |
+      std::views::join | std::views::transform([](auto const& sm) { return std::string_view(sm.first, sm.second); });
+
+  assert(std::ranges::equal(lower, std::to_array<std::string_view>({"e", "l", "l", "o"})));
```
Nitpick, but please `return 0;` from `main`. This is needed in freestanding mode, where `main` is not a special function (so you'd get an error saying you don't return a value from a function that returns `int`).

ldionne reviewed (commented):
Thanks! This is almost good to go, but I am still seeing two review comments I had left which I don't think have been addressed.

JMazurkiewicz committed:
Add `return 0;` in LWG-3698 test

Comment: https://github.com/llvm/llvm-project/pull/66033#discussion_r1409553823

JMazurkiewicz committed:
Revert "Make `as-lvalue.pass.cpp` C++11 friendly"

This reverts commit 0e927818e45d361e3fdc76df9f5938ea291f7bf1.

JMazurkiewicz committed:
Make `as-lvalue.pass.cpp` C++14 friendly

Comment: https://github.com/llvm/llvm-project/pull/66033#discussion_r1409545297

JMazurkiewicz committed:
Test `iterator(Parent& parent, OuterIter outer)` constructor

Comments:
* https://github.com/llvm/llvm-project/pull/66033#discussion_r1396589173
* https://github.com/llvm/llvm-project/pull/66033#discussion_r1396590292 (partially)

JMazurkiewicz committed:
Test `iterator(Parent& parent)` constructor

Comment: https://github.com/llvm/llvm-project/pull/66033#discussion_r1396590292

JMazurkiewicz committed:
Merge branch 'main' into libcxx/ranges/stashing

JMazurkiewicz committed:
Address GCC's complaints

ldionne reviewed (approved):
LGTM, we can merge once CI is green. Thanks!

JMazurkiewicz commented:
Ping @ldionne, CI is green (except for code formatting).
